cache/lru.go

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package cache
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru/simplelru"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type LRUCache struct {
+	c          Cache
+	defaultTTL time.Duration
+	name       string
+
+	mtx sync.Mutex
+	lru *lru.LRU
+
+	requests prometheus.Counter
+	hits     prometheus.Counter
+	items    prometheus.GaugeFunc
+}
+
+type Item struct {
+	Data      []byte
+	ExpiresAt time.Time
+}
+
+// WrapWithLRUCache wraps a given `Cache` c with a LRU cache. The LRU cache will always store items in both caches.
+// However it will only fetch items from the underlying cache if the LRU cache doesn't have the item.
+// Items fetched from the underlying cache will be stored in the LRU cache with a default TTL.
+// The LRU cache will also remove items from the underlying cache if they are expired.
+// The LRU cache is limited in number of items using `lruSize`. This means this cache is not tailored for large items or items that have a big
+// variation in size.
+func WrapWithLRUCache(c Cache, name string, reg prometheus.Registerer, lruSize int, defaultTTL time.Duration) (*LRUCache, error) {
+	lru, err := lru.NewLRU(lruSize, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	cache := &LRUCache{
+		c:          c,
+		lru:        lru,
+		name:       name,
+		defaultTTL: defaultTTL,
+
+		requests: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name:        "cortex_cache_memory_requests_total",
```

Comment:
I get not changing things when doing the initial import but we should probably change this at least, right?

Reply:
For dskit, probably, yes. Do you think we should just do `lru_name_of_metric` instead, make it more configurable with `<your service>_name_of_metric` or just rely on labels etc for that? I lean toward `lru_metric_name` + labels if in dskit. What do you both think?

Reply:
I think just using `lru_metric_name` + labels is good because people can wrap the `reg` being used via `WrapRegistererWithPrefix("cortex_", reg)` if needed

Reply:
> I get not changing things when doing the initial import but we should probably change the cortex_ prefix at least, right?

I overlooked that indeed. 

Done at 6a40d0f3ceb7e7c418de4b22e3dd94472bf97c70. And for the same reason, also got rid of `thanos_` metrics prefix.